### DISCUSSION TITLE
window is not defined

### DISCUSSION THREAD

**Discussion by lisovskiy1337** [vercel/next.js#47028] [Help]
### Summary

Hi guys! Searched a lot about this problem and the most common fix was to add `typeof window !== "undefined"`, but for some reason it didn't help me. The problem is in redux toolkit slice file. Also I don't know why but the error displays mostly when url is wrong and page not found. Here is example of my code:

```
const isBrowser = () => typeof window !== "undefined";
const LS_favoriteList = JSON.parse(window.localStorage.getItem("favorite") || "[]");

const initialState: IState = {
  favorites: isBrowser() && LS_favoriteList ? LS_favoriteList : []
};
export const favoriteSlice = createSlice({
  name: "favorites",
  initialState,
  reducers: {
    addToFavorite: (state, action: PayloadAction<Mov>) => {
      state.favorites.push(action.payload);
      if (isBrowser()) {
        localStorage.setItem(
          "favorite",
          JSON.stringify(state.favorites)
        );
      }
    },
},
});
```

**Comment by gustavosantaella** [comment]
I have the same error. But I added "use client" in the top file. But not working yet.
```"use client";

import { CButton } from "@coreui/react";
import Icon from "components/Icon";
import { SidebarRightContext } from "contexts/SidebarRigthContext";
import React, { useContext } from "react";

export default function Sidebar({ width = "50%" }) {
  const options = useContext(SidebarRightContext);

  return (
    <>
      {options.open === true && (
        <CButton
          className="mask-overlay cursor-not-allowed"
          onClick={() => options.setOpen(false)}
        />
      )}
      <div
        className="main_sidebar_second shadow col-sm-12 col-md-8 col-xl-12"
        style={{
          width,
          right: options.open ? "0" : -${window?.innerWidth},
        }}
      >
        <div className="container-fluid">
          <Icon
            name="XMarkIcon"
            className="icon"
            onClick={() => {
              options.setOpen(!options.open);
            }}
          />
        </div>
      </div>
    </>
  );
}
```

window is not defined.

**Reply by 634750802** [reply]
`use client` is not `<ClientOnly />`. It's just make the component not to be a react18 server component, and the component still need to be rendered in SSR.

**Reply by gustavosantaella** [reply]
> `use client` is not `<ClientOnly />`. It's just make the component not to be a react18 server component, and the component still need to be rendered in SSR.

I could resolve my issues changing the appDir to pages. Because the pages directory it's client

**Comment by icyJoseph** [comment]
@lisovskiy1337 

The problem is how JS expressions are executed, as you declare them, for example in:

```js
const isBrowser = () => typeof window !== "undefined";
const LS_favoriteList = JSON.parse(window.localStorage.getItem("favorite") || "[]");

const initialState: IState = {
  favorites: isBrowser() && LS_favoriteList ? LS_favoriteList : []
};
```

You are defining, in the second line, an expression, and that means that it is evaluated as you declare it, much like saying:

```js
let x = 2 + 2
```

That'd make 4, so even though you are checking for `isBrowser` further down in the initial state declaration, at line 2 you already checked window.localStorage, without proper guards.

Change it to:

```js
const isBrowser = () => typeof window !== "undefined";
const LS_favoriteList = () => { /* Do your reading the localstorage with fallbacks here */ }

const initialState: IState = {
  favorites: isBrowser() ? LS_favoriteList() : []
};
```

**Comment by icyJoseph** [comment]
@gustavosantaella 

Client components are rendered on the server too. These so called "client components" are the same as, the components found in any React app up to these days.

That means that components, including pages, within the pages directory also behave this way. I don't understand how changes from appDir to pages, could've had any impact in your problem, though I haven't really read through your example. Are you sure it wasn't something else that fixed it? Like `options.open` being initially `true`?

**Comment by naz3eh** [comment]
ran into the same error and am struggling to solve it for a long.
